src/compile/facet.ts

```diff
@@ -121,7 +121,7 @@ export class FacetModel extends ModelWithField {
     parseFacetHeaders(this);
   }
 
-  public assembleSelectionTopLevelSignals(signals: NewSignal[]): NewSignal[] {
+  public assembleSelectionTopLevelSignals(signals: readonly NewSignal[]): readonly NewSignal[] {
     return this.child.assembleSelectionTopLevelSignals(signals);
```

Comment:
I'm surprised this works because `assembleSelectionTopLevelSignals` is not pure by design, and definitely makes changes to the `signals` array. As a result, I'm a little worried whether `readonly` poses a danger in being a red herring (e.g., we think functions are pure when they are not).

Reply:
Removed